runtime/src/qio/qio_popen.c

```diff
@@ -0,0 +1,494 @@
+ * limitations under the License.
+ */
+
+#include "sys_basic.h"
+
+#ifndef CHPL_RT_UNIT_TEST
+#include "chplrt.h"
+#endif
+
+#include "qio.h"
+#include "sys.h"
+
+#include "qio_popen.h"
+
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <spawn.h>
+
+// current environment variables
+extern char** environ;
+
+/* Set up file actions for posix_spawn.
+   *std__fd is FD_FORWARD, FD_CLOSE, FD_PIPE etc or a file descriptor #
+   pipe[2] is the pipe created for FD_PIPE
+   childfd is 0 (for stdin) 1 (for stdout) 2 (for stderr)
+   input_to_child is 1 for stdin, 0 otherwise
+   *hasactions is set on output if we added any file actions
+   */
+static qioerr setup_actions(
+    posix_spawn_file_actions_t * actions,
+    int* std__fd, int pipe[2], int childfd,
+    int input_to_child,
+    bool * hasactions)
+{
+  int pipe_parent_end;
+  int pipe_child_end;
+  int rc;
+
+  assert(childfd == 0 || childfd == 1 || childfd == 2);
+
+  // for stdin, the parent end is pipe[1], child end is pipe[0];
+  // for everything else, the parent end is pipe[0], child end is pipe[1].
+  if( input_to_child ) {
+    pipe_parent_end = pipe[1];
+    pipe_child_end = pipe[0];
+  } else {
+    pipe_parent_end = pipe[0];
+    pipe_child_end = pipe[1];
+  }
+
+  if( *std__fd == QIO_FD_FORWARD ) {
+    // Do nothing. Assume file descriptor childfd does not have close-on-exec.
+  } else if( *std__fd == QIO_FD_PIPE ) {
+    // child can't write to the parent end of the pipe. 
+    rc = posix_spawn_file_actions_addclose(actions, pipe_parent_end);
+    if( rc ) return qio_int_to_err(errno);
+
+    // child needs to use its end of the pipe as fd childfd.
+    rc = posix_spawn_file_actions_adddup2(actions, pipe_child_end, childfd);
+    if( rc ) return qio_int_to_err(errno);
+
+    // then close the pipe we dup'd since it is now known by another
+    // name (e.g. fd 0).
+    posix_spawn_file_actions_addclose(actions, pipe_child_end);
+    *hasactions = true;
+  } else if( *std__fd == QIO_FD_CLOSE ) {
+    // close stdin.
+    rc = posix_spawn_file_actions_addclose(actions, childfd);
+    if( rc ) return qio_int_to_err(errno);
+    *hasactions = true;
+  } else if( *std__fd == QIO_FD_TO_STDOUT ) {
+    // Do nothing.
+    assert(0);
+  } else {
+    // Use a given file descriptor for childfd (e.g. stdin).
+    rc = posix_spawn_file_actions_adddup2(actions, *std__fd, childfd);
+    if( rc ) return qio_int_to_err(errno);
+    // then close the pipe we dup'd
+    rc = posix_spawn_file_actions_addclose(actions, *std__fd);
+    if( rc ) return qio_int_to_err(errno);
+    *hasactions = true;
+  }
+
+  return 0;
+}
+
+/* Each of stdin/stdout/stderr take in a pointer to a file descriptor.
+   *ptr should be:
+    -1 -> use the existing stdin/stdout
+    -2 -> close the file descriptor
+    -3 -> create a pipe and return the parent's end
+    >0 -> use this file discriptor
+   When this function returns, any file descriptors with -1
```

Comment:
I'll update the comment - it looks like the code unconditionally sets *stdin_fd etc if a pipe is used (as with -3).